File: x/go/gorp/update.go
// Copyright 2026 Synnax Labs, Inc.
//
// Use of this software is governed by the Business Source License included in the file
// licenses/BSL.txt.
//
// As of the Change Date specified in that file, in accordance with the Business Source
// License, use of this software will be governed by the Apache License, Version 2.0,
// included in the file licenses/APL.txt.

package gorp

import (
	"context"

	"github.com/synnaxlabs/x/errors"
	"github.com/synnaxlabs/x/query"
)

// Update is a query that updates Entries in the DB.
type Update[K Key, E Entry[K]] struct {
	retrieve Retrieve[K, E]
	changes  changes[K, E]
	// indexes mirrors Create.indexes — propagated by Table.NewUpdate so the
	// writer built in Exec stages mutations against the Table's secondary
	// indexes.
	indexes []Index[K, E]
}

// NewUpdate opens a new Update query.
func NewUpdate[K Key, E Entry[K]]() Update[K, E] {
	return Update[K, E]{retrieve: NewRetrieve[K, E]()}
}

// Where adds the provided filter to the query. To update by primary key,
// compose MatchKeys into the filter (e.g. u.Where(MatchKeys(1, 2, 3))).
func (u Update[K, E]) Where(filter Filter[K, E]) Update[K, E] {
	u.retrieve = u.retrieve.Where(filter)
	return u
}

func (u Update[K, E]) Change(f func(Context, E) E) Update[K, E] {
	return u.ChangeErr(func(ctx Context, e E) (E, error) { return f(ctx, e), nil })
}

func (u Update[K, E]) ChangeErr(f func(Context, E) (E, error)) Update[K, E] {
	u.changes = append(u.changes, f)
	return u
}

func (u Update[K, E]) Exec(ctx context.Context, tx Tx) (err error) {
	checkForNilTx("update.Exec", tx)
	var entries []E
	if err = u.retrieve.Entries(&entries).Exec(ctx, tx); err != nil {
		return err
	}
	if len(u.changes) == 0 {
		return errors.Wrap(query.ErrInvalidParameters, "[gorp] - update query must specify at least one change function")
	}
	for i, e := range entries {
		if entries[i], err = u.changes.exec(Context{Context: ctx, Tx: tx}, e); err != nil {
			return err
		}
	}
	return wrapWriter[K, E](tx, u.retrieve.keyPrefix, u.indexes).Set(ctx, entries...)
}

type ChangeFunc[K Key, E Entry[K]] = func(Context, E) (E, error)

type changes[K Key, E Entry[K]] []ChangeFunc[K, E]

func (c changes[K, E]) exec(ctx Context, entry E) (o E, err error) {
	for _, change := range c {
		if o, err = change(ctx, entry); err != nil {
			return
		}
	}
	return
}

File: x/ts/src/array/update.ts
// Copyright 2026 Synnax Labs, Inc.
//
// Use of this software is governed by the Business Source License included in the file
// licenses/BSL.txt.
//
// As of the Change Date specified in that file, in accordance with the Business Source
// License, use of this software will be governed by the Apache License, Version 2.0,
// included in the file licenses/APL.txt.

import { type record } from "@/record";

export const upsert = <V>(prev: V[] | null | undefined, changes: V | V[]): V[] => {
  if (Array.isArray(changes))
    return changes.reduce((acc, v) => upsert(acc, v), prev ?? []);
  if (prev == null) return [changes];
  const idx = prev.findIndex((i) => i == changes);
  const next = [...prev];
  if (idx === -1) next.push(changes);
  else next[idx] = changes;
  return next;
};

export interface Remove {
  <V>(prev: V[], deletes: V | V[]): V[];
  <V>(prev: V[] | undefined, deletes: V | V[]): V[] | undefined;
}

export const remove = (<V>(
  prev: V[] | undefined,
  deletes: V | V[],
): V[] | undefined => {
  if (prev == null) return undefined;
  if (Array.isArray(deletes)) return prev.filter((v) => !deletes.includes(v));
  return prev.filter((v) => v != deletes);
}) as Remove;

export const upsertKeyed = <K extends record.Key, E extends record.Keyed<K>>(
  prev: E[] | null | undefined,
  changes: E | E[],
): E[] => {
  if (Array.isArray(changes))
    return changes.reduce((acc, t) => upsertKeyed(acc, t), prev ?? []);
  if (prev == null) return [changes];
  const idx = prev.findIndex((i) => i.key === changes.key);
  const next = [...prev];
  if (idx === -1) next.push(changes);
  else next[idx] = changes;
  return next;
};

export interface RemoveKeyed {
  <K extends record.Key, E extends record.Keyed<K>>(prev: E[], keys: K | K[]): E[];
  <K extends record.Key, E extends record.Keyed<K>>(
    prev: E[] | undefined,
    keys: K | K[],
  ): E[] | undefined;
}

export const removeKeyed = (<K extends record.Key, E extends record.Keyed<K>>(
  prev: E[] | undefined,
  key: K | K[],
): E[] | undefined => {
  if (prev == null) return undefined;
  if (Array.isArray(key)) return prev.filter((i) => !key.includes(i.key));
  return prev.filter((i) => i.key !== key);
}) as RemoveKeyed;
